google/cloud/storage/oauth2/google_credentials.h

```diff
@@ -62,21 +63,61 @@ CreateAuthorizedUserCredentialsFromJsonFilePath(std::string const&);
  * with Cloud Storage client libraries.
  */
 StatusOr<std::shared_ptr<Credentials>>
-CreateAuthorizedUserCredentialsFromJsonContents(std::string const&);
+CreateAuthorizedUserCredentialsFromJsonContents(std::string const& contents);
 
-/// Creates a ServiceAccountCredentials rom a JSON file at the specified path.
+/**
+ * Creates a ServiceAccountCredentials from a JSON file at the specified path.
+ *
+ * @param path the path to the file containing service account JSON credentials.
+ * @param scopes the scopes to request during the authorization grant. If
+ *     omitted or an empty set is provided, the cloud-platform scope,
+ *     defined by `GoogleOAuthScopeCloudPlatform()`, is used.
+ * @param subject for domain-wide delegation; the email address of the user for
+ *     which to request delegated access. If omitted or an empty string is
+ *     provided, no "subject" attribute is included in the authorization grant.
+ *
+ * @see https://developers.google.com/identity/protocols/OAuth2ServiceAccount
+ *     for more information about domain-wide delegation.
+ *
+ * @see https://developers.google.com/identity/protocols/googlescopes for a list
+ *     of OAuth 2.0 scopes used with Google APIs.
+ */
 StatusOr<std::shared_ptr<Credentials>>
-CreateServiceAccountCredentialsFromJsonFilePath(std::string const&);
+CreateServiceAccountCredentialsFromJsonFilePath(
+    std::string const& path,
+    std::set<std::string> const& scopes = std::set<std::string>(),
+    std::string const& subject = "");
```

Comment:
It looks like both the scopes and subject arguments are optional. Rather than using the "empty" object (both this string and the scopes set) as the sentinel indicating to omit or use the default, what do you think about using std::optional instead? This would make it clearer in the signature that these arguments are indeed optional. It would also avoid carving out an otherwise valid value from within the domain of the argument type (e.g., carving out the empty string from the set of all strings) to stand in for this "null opt" value. In other words, if a caller supplies a string for the subject param -- any string -- this code should use it. Even including the empty string (probably). 

I realize it may seem strange to allow the caller to legitimately specify an empty subject, but this code is already allowing the caller to specify a subject with just a single space, or any other conceivable string. Why would then the empty string be special? 

Then *IF* we change to using std::optional to represent "optional" parameters, we might want to consider not defaulting the parameters. That is, requiring the callers to pass `std::nullopt` if they explicitly want to use the default value. This way it's clear at the call site that there are additional parameters that could be passed but the caller is choosing to not set them.

I realize this would be a breaking change. So if that's a problem, we can keep the default arguments.

... thoughts?

Reply:
SGTM. I do prefer overloads to defaulted params in general. FWIW, one reason for this preference is that defaulted params don't change a function's actual signature so sometimes things that need the actual signature of the fuction (like a function type or a function pointer) may look weird compared to what is actually called.